### ISSUE TITLE
compiler: DSA compressor ops — Ca/Cb entry compression for sparse attention

### ISSUE THREAD

**Issue by xzyaoi** [opentela-ai/vkernels#96]
## Context
DeepSeek Sparse Attention (floe `deepseek_v4/forward.py::_BaseCompressor`/`Compressor`, v41 `Compressor`): every m=4 tokens compress into one cached entry (two overlapping Ca/Cb series for CSA; m=128 for HCA), entries rotated at emission so decode only rotates the query.

## Semantics (decode step)
- when `p_row % m == m-1`: append one compressed entry (small projections + rope) to the per-layer compressor state pool — gated by a **per-row** condition (needs #93 per-row lengths; the append task runs masked on rows not at a boundary)
- attention then runs over [sliding window ∪ compressed entries ∪ (selected top-k entries — see indexer issue)]

## Work
- `compressor_append` op (state-pool RMW, `cache_append` pattern), `compressor_scores`/`values` fused with window attention or as separate families — profiling decides
- reference bodies from the floe eager path

## Validation
entry-level equality vs eager compressor state after N tokens; decode logits parity.

Labels: arch:ops, model:deepseek-v4

## Implementation sketch

**Backend**: Triton. `compressor_append` = `cache_append`-shaped state task (one per layer, masked on rows whose `p_row % m == m−1`); compressed-entry reads fold into the MLA scores/values tasks (#95) via slot indirection (#94) — no separate gather phase.

**Arithmetic** (emission at a boundary; CSA m=4, HCA m=128; fp32 softmax stats):
```
w_t  = softmax(gate[window_t])                  # fp32 over the m-token window
e    = rms_norm(Σ_t w_t · kv_t) · rms_weight     # weighted fold of the shared latent
e    = rope_slice(e, cos, sin at emission position)  # rotated ONCE at emission…
# …so decode rotates only the query (entries stay unrotated in the pool)
```
CSA keeps **two overlapping series** (Ca prior window, Cb current, width 2r stride r) — at decode: append to Cb; at the r-boundary, Cb becomes the new Ca and Cb restarts. `block_bias` per entry (from the indexer, #97) joins the scores mask.

**Dtypes**: entry pool bf16 `[S/m, head_dim]` per layer per row; in-flight raw window + gates bf16/fp32 in the compressor state pool; all folds fp32-accumulated.

**Comment by xzyaoi** [comment]
ISSUE 96 REPORT — compiler lane complete, branch `issue-96-dsa-compressor` pushed, PR #125 open (stacks #97's branch — merge #119 first).

**Delivered (4 commits on top of the #97 stack)**
- `e90c3e8` `OP_COMPRESSOR_APPEND` + capture recorder: emission/rotation/block_bias numerical contract; post-append views (cache_append §4.3); per-row position form (#93); conservative per-row slab write regions.
- `33b21b3` lowering — domain (B, L), one task per (batch, layer), params m/r/eps/position — + reference-exec body (fp64-exact emission, boundary no-op, slot ping-pong) + POSITION_CONSUMERS admission.
- `f0373bc` `_t_compressor_append` Triton template + `compressor_append_task` codegen registration.
- `58745b2` bare-env CPU oracle suite.

**Semantics implemented**: boundary-masked per-row emission (`p[b] % m == m-1`); fp32 softmax-over-gates → weighted latent fold → rms_norm → rotate_half rope **once at the emitting row's position**; two-series Ca/Cb pool with slot-role ping-pong at `cb_len == r//m`; [Ca ∪ Cb] window = 2R consecutive global entries sliding by R per rotation (width 2r stride r); block_bias alignment contract consumed as documented (#97 producer of record; no mismatch surfaced).

**Verified here (bare numpy env)**: suite `12 passed, 1 skipped` — ragged fp64-mirror walk tile-exact; boundary-mask byte-identity; rotation swap/restart; overlap geometry; rope-once identity (score q_rot·e_stored == q·e_raw, rtol 1e-9); NaN canaries; chained-append RAW/WAR/WAW on series state; template algorithmic mirror; task grid B·L. Full lane regression `109 passed, 10 skipped`.

**Attested, not verified**: Triton template CUDA run (sweep-owned); floe `_BaseCompressor` eager parity (import-gated).

**Follow-ups flagged**: (1) `rope_slice` implemented full-width — partial-slice width is a one-parameter change if V4's compressor ropes a subset; (2) in-flight window maintenance + scores/values fold-in are #95's integration surface.

**Comment by xzyaoi** [comment]
Landed via #125 (merged): compressor_append (two-series Ca/Cb entry emission). OP_COMPRESSOR_APPEND is present in operator_ir.py on main. Closing.
